src/diagnostics.rs

```diff
@@ -0,0 +1,569 @@
+use crate::SourceRanged;
+use crate::SourceTextInfo;
+
+pub enum DiagnosticLevel {
+  Error,
+  Warning,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct DiagnosticSourceRange {
+  pub start: DiagnosticSourcePos,
+  pub end: DiagnosticSourcePos,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum DiagnosticSourcePos {
+  SourcePos(SourcePos),
+  ByteIndex(usize),
+  LineAndCol {
+    // 0-indexed line number in bytes
+    line: usize,
+    // 0-indexed column number in bytes
+    column: usize,
+  },
+}
+
+impl DiagnosticSourcePos {
+  fn pos(&self, source: &SourceTextInfo) -> SourcePos {
+    match self {
+      DiagnosticSourcePos::SourcePos(pos) => *pos,
+      DiagnosticSourcePos::ByteIndex(index) => source.range().start() + *index,
+      DiagnosticSourcePos::LineAndCol { line, column } => {
+        source.line_start(*line) + *column
+      }
+    }
+  }
+}
+
+#[derive(Clone, Debug)]
+pub enum DiagnosticLocation<'a> {
+  /// The diagnostic is relevant to a specific path.
+  Path { path: PathBuf },
+  /// The diagnostic is relevant to an entire module.
+  Module {
+    /// The specifier of the module that contains the diagnostic.
+    specifier: Cow<'a, ModuleSpecifier>,
+  },
+  /// The diagnostic is relevant to a specific position in a module.
+  ///
+  /// This variant will get the relevant `SouceTextInfo` from the cache using
+  /// the given specifier, and will then calculate the line and column numbers
+  /// from the given `SourcePos`.
+  ModulePosition {
+    /// The specifier of the module that contains the diagnostic.
+    specifier: Cow<'a, ModuleSpecifier>,
+    /// The source position of the diagnostic.
+    source_pos: DiagnosticSourcePos,
+    text_info: Cow<'a, SourceTextInfo>,
+  },
+}
+
+impl<'a> DiagnosticLocation<'a> {
+  /// Return the line and column number of the diagnostic.
+  ///
+  /// The line number is 1-indexed.
+  ///
+  /// The column number is 1-indexed. This is the number of UTF-16 code units
+  /// from the start of the line to the diagnostic.
+  /// Why UTF-16 code units? Because that's what VS Code understands, and
+  /// everyone uses VS Code. :)
+  fn position(&self) -> Option<(usize, usize)> {
+    match self {
+      DiagnosticLocation::Path { .. } => None,
+      DiagnosticLocation::Module { .. } => None,
+      DiagnosticLocation::ModulePosition {
+        specifier: _specifier,
+        source_pos,
+        text_info,
+      } => {
+        let pos = source_pos.pos(text_info);
+        let line_index = text_info.line_index(pos);
+        let line_start_pos = text_info.line_start(line_index);
+        // todo(dsherret): fix in text_lines
+        let content =
+          text_info.range_text(&SourceRange::new(line_start_pos, pos));
+        let line = line_index + 1;
+        let column = content.encode_utf16().count() + 1;
+        Some((line, column))
+      }
+    }
+  }
+}
+
+pub struct DiagnosticSnippet<'a> {
+  /// The source text for this snippet. The
+  pub source: Cow<'a, crate::SourceTextInfo>,
```

Comment:
Forgot to point out this change I did make which was to only accept a `SourceTextInfo` here instead of also potentially a `ModuleSpecifier`